### ISSUE TITLE
python-systemd is not in jessie-saltstack

### ISSUE THREAD

**Issue by babilen** [saltstack/salt-pack#25]
The current `salt-master` depends on `python-systemd | sysvinit-core` which causes the installation of sysvinit-core during the salt-master upgrade, because `python-systemd` is not available in jessie. This forces a switch from SystemD to SysVinit, which is surely neither intentional nor fortunate.

**Comment by dmurphy18** [comment]
@babilen pulled this change from the debian changes for salt_2015_8_3+ds~3.
I can remove that line.  But I wonder if the changes in debian's salt_2015_8_3+ds~3 correct, or did I misunderstand something in their changes.

**Comment by babilen** [comment]
The debian changes are fine as they target unstable in which `python-systemd` is available. If you build for jessie you'll have to backport that package however.

**Comment by dmurphy18** [comment]
@babilen Thanks for catching that, I missed 'unstable' and 'python-systemd'.  I shall remove the line from salt-master and it should be available in the next point release which will be very soon.

If this is satisfactory, can this issue be closed.

**Comment by babilen** [comment]
I think that the right approach would be to actually backport `python-systemd`.

**Comment by dmurphy18** [comment]
Unfortunately I am very constrained by time on my task list.
It would be about a month before I could get to that backport, but I'll look into doing that when I can.

**Comment by babilen** [comment]
Okay, the problem is that this causes problems in `process.py`, `daemons.py` and probably a few other places where `python-systemd` is used. I'll try to get `.5` packaged in the coming days and to backports `python-systemd` and get it into the community repository.

**Comment by dmurphy18** [comment]
@babilen  I have it removed in the latest point release which is built and about to under go testing.  The point release is immanent due to a bug in the state compiler caught 10 minutes after announcing. Hence the time constraints, but the backport would be great to have. Just wanted to make you fully informed, that the need for the backport will not be so great.

**Comment by babilen** [comment]
@dmurphy18 Yeah, I've been around yesterday when the `watch_in` regression was discussed. Good that `.5` is being cut immediately, but the python-systemd situation should still be sorted out.

**Comment by dmurphy18** [comment]
@babilen  I agree, I shall look into it, when I get some time, but won't be immediately due to .5 priorities.  i shall try for the point release after.

**Comment by babilen** [comment]
The problem will be that https://github.com/saltstack/salt/blob/2015.8/pkg/salt-master.service#L7 will cause systemd to wait for a callback which is done in https://github.com/saltstack/salt/blob/2015.8/salt/utils/process.py#L35 -- If `python-systemd` is not installed this callback will never be send as `systemd.daemon` cannot be imported.

**Comment by dmurphy18** [comment]
@babilen Thanks for that info, 2015.8.4 has a work-around for that issue. I shall adjust for this in following point releases.  It was on my list to investigate the issue, but point release 2015.8.5 occurred.

**Comment by babilen** [comment]
@dmurphy18 Ah, great. Out of curiosity: What is the workaround?

**Comment by dmurphy18** [comment]
@babilen use salt-master.service from debian directory and replace notify with simple.  Pragmatic fix to get under a deadline for release.  i was also trying to get salt-master.service to read environment and use that for Restart, but ran out of time trying to resolve correct syntax to utilize env. value, which may not be possible with Debian (saw something to that effect when searching to resolve the issue)

**Comment by babilen** [comment]
Yeah, replacing `notify` is exactly what I had in mind too, but that might not be necessary in the community repositories when a backport is available. Once that is done it should hopefully not be too tricky to integrate a backport into salt-pack.

**Comment by jdelic** [comment]
any chance of getting this ticket a higher priority? Coming from #30696 which is rated *critical* and which seems to be blocked by this ticket, that would seem helpful.

And if I might ask: Are there particular problems about backporting python-systemd or its dependencies to Debian Jessie? I assume that providing that package in a "third-party repo" on our end would be a viable workaround for #30696, right?

**Comment by babilen** [comment]
@jdelic I'm working on 2015.8.5 packages for `jessie-saltstack` (i.e. https://debian.saltstack.com) after @bdrung and I finished the packages for unstable today. Building that version on jessie requires a number of backports (python-systemd, sphinx, python-alabaster and python-pygments) however, which are fortunately relatively straightforward.

Once we are done with that I think that it would be great to get these backports integrated into salt-pkg, which should hopefully not be too hard.

**Comment by jdelic** [comment]
> Building that version on jessie requires a number of backports (python-systemd, sphinx, python-alabaster and python-pygments) however, which are fortunately relatively straightforward.

Hmm, I just built the `testing` source package from the official Debian repo under Jessie using `apt-get --build source python-systemd` and it built without any of those dependencies, seemed to install fine and allowed me to switch back to `systemd-sysv`. The dependency list reads: `Depends: python (<< 2.8), python (>= 2.7~), python:any (>= 2.7.5-5~), libc6 (>= 2.17), libsystemd0 (>= 214)`. Am I missing something?

**Comment by babilen** [comment]
@jdelic First and foremost: testing has `2015.8.3+ds-3` rather than the `2015.8.5+ds-1` version we uploaded today, but that is not responsible for the dependencies in this case. The Build-Depends include, among other things, `python-sphinx (>= 1.3)` which is not in `jessie` and therefore can't be satisfied.

You obviously had `jessie-backports` enabled which allows you to pull in a number of dependencies that would not normally have been available.

All that being said, I'm done with the packages for jessie and will test them tomorrow.

**Comment by dmurphy18** [comment]
@babilen @jdelic FYI in the point release 2015.8.5 which is immanent, I pulled the 'python-sphinx (>= 1.3)', it is 'python-sphinx (>= 1.0)'.  doc's are now pulled from the pre-built docs in the salt tarball, e.g. salt-2015.8.5.tar.gz

**Comment by dmurphy18** [comment]
@babilen with the release of 2015.8.5, the offending line was removed, i.,e. the dependency of salt-master on python-systemd | sysvinit-core.

Can this issues be closed ?

**Comment by babilen** [comment]
Well, that fixes https://github.com/saltstack/salt/issues/30696, but I still think that it would be good to backport `python-systemd` and allow users to take advantage of `Type=notify`, so I wouldn't close it just yet.

**Comment by dmurphy18** [comment]
@babilen that is another issue. I think it would be better to have a different task for that, and close this, since the issue stated here is fixed.

**Comment by dmurphy18** [comment]
@babilen I am closing to close this, if you feel strongly different, I will reopen it.
It would be better to have the backport issue handled by a new issue.

**Comment by babilen** [comment]
@dmurphy: python-systemd has been available in jessie-backports for quite a while now and the workaround is no longer needed. I would be quite happy to see "Type=notify" back in the unit files :)

**Comment by dmurphy18** [comment]
@babilen I shall take a look, at that and other changes made on the Debian side for 2016.3.0 recently, saw it go into unstable. You had mentioned the need for the notify message to be processed, wonder if you have that file information to hand, if not, I can grep through the old emails.

**Comment by dmurphy18** [comment]
@babilen never mind on the info, I scrolled up

**Comment by dmurphy18** [comment]
@babilen Investigating into this has shown an issue with systemd support which shall be addressed by https://github.com/saltstack/salt/issues/33803. Basically, salt's use of systemd notify has to be implemented for more than just the salt-master starting.
